From 089dea799f471728ca800fb976e760ced8e9419b Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:06:57 -0400
Subject: [powerpc] eeh: Enable PCI_COMMAND_MASTER for PCI bridges

Message-id: <1410545655-205645-189-git-send-email-dzickus@redhat.com>
Patchwork-id: 94191
O-Subject: [RHEL7 PATCH 188/626] powerpc/eeh: Enable PCI_COMMAND_MASTER for PCI bridges
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit bf898ec5cbd33be11147743bee27b66b10cb2f85
Author: Gavin Shan <shangw@linux.vnet.ibm.com>
Date:   Tue Nov 12 14:49:21 2013 +0800

    powerpc/eeh: Enable PCI_COMMAND_MASTER for PCI bridges

    On PHB3, we will fail to fetch IODA tables without PCI_COMMAND_MASTER
    on PCI bridges. According to one experiment I had, the MSIx interrupts
    didn't raise from the adapter without the bit applied to all upstream
    PCI bridges including root port of the adapter. The patch forces to
    have that bit enabled accordingly.

    Signed-off-by: Gavin Shan <shangw@linux.vnet.ibm.com>
    CC: <stable@vger.kernel.org>
    Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/kernel/eeh.c b/arch/powerpc/kernel/eeh.c
index 3759df4..52e1520 100644
--- a/arch/powerpc/kernel/eeh.c
+++ b/arch/powerpc/kernel/eeh.c
@@ -687,6 +687,15 @@ void eeh_save_bars(struct eeh_dev *edev)
 
  for (i = 0; i < 16; i++)
   eeh_ops->read_config(dn, i * 4, 4, &edev->config_space[i]);
+
+ /*
+  * For PCI bridges including root port, we need enable bus
+  * master explicitly. Otherwise, it can't fetch IODA table
+  * entries correctly. So we cache the bit in advance so that
+  * we can restore it after reset, either PHB range or PE range.
+  */
+ if (edev->mode & EEH_DEV_BRIDGE)
+  edev->config_space[1] |= PCI_COMMAND_MASTER;
 }
 
 /**
-- 
1.7.1